## Введение в высшую математику
### Практическое задание №6

#### 1. Решите линейную систему: A*X=B

In [2]:
import numpy as np
import matplotlib.pyplot as plt

In [3]:
A = np.array([[1, 2, 3], [4, 0, 6], [7, 8, 9]])
B = np.array([12, 2, 1])
np.linalg.solve(A, B)

array([-9.2       ,  0.9       ,  6.46666667])

#### 2. Найдите псевдорешение следующей СЛАУ: 
x + 2y – z = 1

3x – 4y = 7

8x – 5y + 2z = 12

2x – 5z = 7

11x +4y – 7z = 15


##### Так как это СЛАУ переопределенное, то поиск псевдорешения сводится к поиску вектора,который бы минимизировал норму невязки системы этих уровнений:

In [4]:
A = np.array([[1, 2, -1], [2, -4, 0], [8, -5, 2], [2, -5, 0], [11, 4, -7]])
B = np.array([1, 7, 12, 7, 15])

# Псевдорешение методом наименьших квадратов (решение минимизирующее длину вектора невязки)
np.linalg.lstsq(A, B)

<ipython-input-4-090aa444a13b>:4: FutureWarning: `rcond` parameter will change to the default of machine precision times ``max(M, N)`` where M and N are the input matrix dimensions.
To use the future default and silence this warning we advise to pass `rcond=None`, to keep using the old, explicitly pass `rcond=-1`.
  np.linalg.lstsq(A, B)


(array([ 1.11788888, -1.01092102, -0.98384686]),
 array([1.47060045]),
 3,
 array([14.75062949, 10.40376988,  2.86015769]))

In [7]:
# Проверка через подстановку всевдокорней:

np.dot(A, [1.11788888, -1.01092102, -0.98384686])

array([ 0.0798937 ,  6.27946184, 12.03002242,  7.29038286, 15.14002162])

In [8]:
# Посчитаем невязку:

np.dot(A, [1.11788888, -1.01092102, -0.98384686]) - B

array([-0.9201063 , -0.72053816,  0.03002242,  0.29038286,  0.14002162])

In [10]:
# Функция нормы невязки: 

def Q(x, y, z):
    return ((np.linalg.norm(np.dot(A, [x, y, z]) - B))**2)
Q(1.11788888, -1.01092102, -0.98384686)

1.470600448467736

#### 3. Сколько решений имеет линейная система:


In [20]:
A = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
B = np.array([12, 2, 1])

np.linalg.solve(A, B)

array([ 2.83726777e+16, -5.67453553e+16,  2.83726777e+16])

In [14]:
# Проверка:

np.dot(A, [2.83726777e+16, -5.67453553e+16,  2.83726777e+16])

array([2.e+08, 5.e+08, 8.e+08])

In [21]:
np.linalg.matrix_rank(A, 0.0001)

2

#### Ответ: система имеет 2 решения: x1 ~ 2.84, x2 ~ -5.67

#### 4. Вычислите LU-разложение матрицы:

In [37]:
import numpy as np
import matplotlib.pyplot as plt
import scipy
import scipy.linalg

In [41]:
A = np.array([[1, 2, 3], [2, 16, 21], [4, 28,73]])
P, L, U = scipy.linalg.lu(A)

print(P)
print(L)
print(U)

print(np.dot(P, A) - np.dot(P, A))

[[0. 1. 0.]
 [0. 0. 1.]
 [1. 0. 0.]]
[[ 1.    0.    0.  ]
 [ 0.25  1.    0.  ]
 [ 0.5  -0.4   1.  ]]
[[  4.    28.    73.  ]
 [  0.    -5.   -15.25]
 [  0.     0.   -21.6 ]]
[[0. 0. 0.]
 [0. 0. 0.]
 [0. 0. 0.]]


#### 5. Найдите нормальное псевдорешение недоопределенной системы:
##### x + 2y – z = 1  
##### 8x – 5y + 2z = 12

#### Для этого определите функцию Q(x,y,z), равную норме решения, и найдите ее минимум.


In [45]:
A = ([[1, 2, -1], [8, -5, 2]])
B = ([1, 12])

np.linalg.lstsq(A,B)

<ipython-input-45-98c82a7efca0>:4: FutureWarning: `rcond` parameter will change to the default of machine precision times ``max(M, N)`` where M and N are the input matrix dimensions.
To use the future default and silence this warning we advise to pass `rcond=None`, to keep using the old, explicitly pass `rcond=-1`.
  np.linalg.lstsq(A,B)


(array([ 1.38191882, -0.18081181,  0.0202952 ]),
 array([], dtype=float64),
 2,
 array([9.65316119, 2.41173777]))

In [49]:
# Функция нормы решения (квадраты нормы): 

def Q(x, y, z):
    return (x**2 + y**2 + z**2)
Q(1.38191882, -0.18081181,  0.0202952)

1.942804430848709

#### Ответ: псевдорешение данной недоопределенной системы находится в точке с координатами (1.38191882, -0.18081181,  0.0202952)

### 6. Найдите одно из псевдорешений вырожденной системы:

In [50]:
A = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
B = np.array([2, 5, 11])

Q, R = np.linalg.qr(A)

print(A)
print(Q)
print(R)

[[1 2 3]
 [4 5 6]
 [7 8 9]]
[[-0.12309149  0.90453403  0.40824829]
 [-0.49236596  0.30151134 -0.81649658]
 [-0.86164044 -0.30151134  0.40824829]]
[[-8.12403840e+00 -9.60113630e+00 -1.10782342e+01]
 [ 0.00000000e+00  9.04534034e-01  1.80906807e+00]
 [ 0.00000000e+00  0.00000000e+00 -1.11164740e-15]]


In [51]:
print(np.dot(Q, R))
print(np.dot(np.transpose(Q), Q))

[[1. 2. 3.]
 [4. 5. 6.]
 [7. 8. 9.]]
[[ 1.00000000e+00 -5.26517217e-16 -2.55176183e-16]
 [-5.26517217e-16  1.00000000e+00  3.37757775e-16]
 [-2.55176183e-16  3.37757775e-16  1.00000000e+00]]


In [52]:
R1 = R[:2, :2]
R1

array([[-8.1240384 , -9.6011363 ],
       [ 0.        ,  0.90453403]])

In [53]:
B1 = np.dot(np.transpose(Q), B) [:2]
B1

array([-1.21860576e+01,  8.54871729e-15])

In [54]:
X1 = np.linalg.solve(R1, B1)
X1

array([1.50000000e+00, 9.45096256e-15])

#### Ответ: Одним из псевдорешений является 1.5 и 9.45 (при Х2 = 0)

In [61]:
# Посчитаем норму: 

X = np.append(X1, 0)
print(X)
np.linalg.norm(X)

[1.50000000e+00 9.45096256e-15 0.00000000e+00]


1.4999999999999893

In [62]:
# Посчитаем невязку линейной системы:

np.linalg.norm(np.dot(A, X) - B)

1.2247448713915885

#### Для нахождения нормального псевдорешения непонятно как найти минимальную норму?

In [63]:
# Пока просто в качестве примера как с минимальной нормой [1.5, 1, 0.5] находится нормальное псевдорешение:

X = np.array([1.5, 1, 0.5])
np.linalg.norm(X), np.linalg.norm(np.dot(A, X) - B)

(1.8708286933869707, 15.297058540778355)